### PR TITLE
feat: update RunNewImages based on image diffs

### PR DESCRIPTION
**Pull request by LZiHaN**
This commit sets the RunNewImages variable to be the difference between the current cluster's app image and the desired cluster's app image. This ensures that only the necessary changes are applied when updating the cluster. 

refer to #2832

### CODE REVIEW

File: pkg/apply/utils.go [hunk 1]
```diff
@@ -109,3 +109,24 @@ func getHostArch(sshClient ssh.Interface) func(string) v2.Arch {
 func GetHostArch(sshClient ssh.Interface, ip string) string {
 	return string(getHostArch(sshClient)(ip))
 }
+
+func GetImagesDiff(current, desired []string) []string {
```
Comment (bxy4543) [inline pkg/apply/utils.go:113]: https://github.com/labring/sealos/blob/a02ebcafefbeb31b284ee8786987d7f7c6455721/pkg/utils/strings/strings.go#L174

can use this method to get different images

**Review by bxy4543 — APPROVED** [review]

File: pkg/apply/utils.go (hunk 1 above)
Comment (fengxsong) [inline pkg/apply/utils.go:113]: IMO, we should keep the order of images, sometimes `image 1` might depends on `image 2`

Comment (cuisongliu) [reply pkg/apply/utils.go:113]: > IMO, we should keep the order of images, sometimes `image 1` might depends on `image 2`

Yes, it's better to keep the order while diff.

**Review by cuisongliu — CHANGES_REQUESTED** [review]

File: pkg/apply/utils.go (hunk 1 above)
Comment (LZiHaN) [reply pkg/apply/utils.go:113]: do I need to consider duplicate images in the Clusterfile?

Comment (LZiHaN) [reply pkg/apply/utils.go:113]: do I need to consider the uniqueness of the image for this function as it doesn't seem to have been taken into account?

File: pkg/apply/apply.go [hunk 2]
```diff
@@ -56,10 +56,12 @@ func NewApplierFromFile(path string, args *Args) (applydrivers.Interface, error)
 	}
 	currentCluster := cf.GetCluster()
 
+	runNewImages := GetImagesDiff(currentCluster.Spec.Image, cluster.Spec.Image)
```
Comment (mouuii) [inline pkg/apply/apply.go:59]: before you call GetImagesDiff() ，please use hashfunc to quick compare the spec!

Comment (cuisongliu) [reply pkg/apply/apply.go:59]: > before you call GetImagesDiff() ，please use hashfunc to quick compare the spec!

agree

Comment (LZiHaN) [reply pkg/apply/apply.go]: > > before you call GetImagesDiff() ，please use hashfunc to quick compare the spec!
> 
> agree

done

File: pkg/apply/utils.go (hunk 1 above)
Comment (LZiHaN) [reply pkg/apply/utils.go:115]: > IMO, we should keep the order of images, sometimes `image 1` might depends on `image 2`

done

File: pkg/apply/utils.go [hunk 3]
```diff
@@ -109,3 +111,24 @@ func getHostArch(sshClient ssh.Interface) func(string) v2.Arch {
...
+	return stringsutil.RemoveDuplicate(stringsutil.RemoveStrSlice(desired, current))
+}
+
+func CompareImageSpecHash(currentImages []string, newImages []string) bool {
```
Comment (cuisongliu) [inline pkg/apply/utils.go:119]: hash.ToString replace it.

Comment (LZiHaN) [reply pkg/apply/utils.go:118]: > hash.ToString replace it.

done

**Review by cuisongliu — APPROVED** [review]

**Comment by cuisongliu** [comment]
make format to code

**Review by cuisongliu — CHANGES_REQUESTED** [review]
fix to pass ci

**Comment by cuisongliu** [comment]
```
ok  	github.com/labring/sealos/pkg/env	0.052s	coverage: 74.5% of statements
[188](https://github.com/labring/sealos/actions/runs/4529171618/jobs/7977085728?pr=2856#step:5:189)
--- FAIL: Test_NewApplierFromFile (0.01s)
[189](https://github.com/labring/sealos/actions/runs/4529171618/jobs/7977085728?pr=2856#step:5:190)
    --- FAIL: Test_NewApplierFromFile/success (0.01s)
```
Test_NewApplierFromFile fail

**Comment by LZiHaN** [comment]
> ```
> ok  	github.com/labring/sealos/pkg/env	0.052s	coverage: 74.5% of statements
> [188](https://github.com/labring/sealos/actions/runs/4529171618/jobs/7977085728?pr=2856#step:5:189)
> --- FAIL: Test_NewApplierFromFile (0.01s)
> [189](https://github.com/labring/sealos/actions/runs/4529171618/jobs/7977085728?pr=2856#step:5:190)
>     --- FAIL: Test_NewApplierFromFile/success (0.01s)
> ```
> 
> Test_NewApplierFromFile fail

Through testing, it is found that the error is caused by currentCluster.Spec. I need to modify it.

**Comment by cuisongliu** [comment]
resolve your Conflicting files

**Comment by LZiHaN** [comment]
> resolve your Conflicting files

done

**Review by cuisongliu — APPROVED** [review]

**Pull request merged**